### PR TITLE
Fix #2425: OPALPBA: Reboot after SED-unlocking may hang on UEFI systems

### LINKED ISSUE #2425: OPALPBA: Reboot after unlocking self-encrpyting disks may hang on some UEFI systems
* ReaR version: Relax-and-Recover 2.6 / Git
* OS version: Ubuntu 18.04.4 LTS, Ubuntu 20.04 LTS
* System architecture: x86_64
* Firmware and bootloader: UEFI, Grub 2 without Secure Boot
* Description of the issue:

After unlocking self-encrypting disks via ReaR's TCG Opal-2 PBA, a system is expected to boot into its regular OS. Sequence:
1. System powered on. Self-encrpyted boot disk is locked and shows its PBA boot sector only.
2. The PBA (pre-boot authentication) system boots and asks for a password.
3. The PBA unlocks the boot disk, which now shows its partitions for the 'real' OS.
4. The PBA initiates a reboot.
5. The UEFI firmware runs the bootloader for the real OS (Grub 2 in our example).
6. Grub 2 loads kernel and initramfs.
7. Grub 2 hands over control to the kernel.

Probably due to an UEFI firmware glitch, on an HPE ML10Gen9 server the boot process for the 'real' OS hangs in step 7, probably in the kernel's early initialization phase before console output can be seen.

PR including a workaround for the issue upcoming.

### PR DESCRIPTION
**Pull request by OliverO2**
##### Pull Request Details:

* Type: **Bug Fix**

* Impact: **Normal**

* Reference to related issue (URL): #2425

* How was this pull request tested? On Ubuntu 20.04 LTS on two UEFI systems (one HPE ML10Gen9 server with problematic firmware, one other)

* Brief description of the changes in this pull request:
    * OPALPBA: Add and remove a UEFI 'next boot' entry (essentially a no-op)

### CODE REVIEW

**Comment by jsmeix** [comment]
@rear/contributors 
because this change affects only `OUTPUT=RAWDISK`
and because it was tested by @OliverO2
I would like to merge it tomorrow morning
if there are no objections.

File: usr/share/rear/skel/default/etc/scripts/unlock-opal-disks [hunk 1]
```diff
@@ -201,6 +201,20 @@ while (( attempt < max_authentications )); do
         else
             display_message "$unlocked_device_count of $device_count disks unlocked, rebooting..."
         fi
+        if type -p efibootmgr &>/dev/null; then
+            # Workaround for EFI firmware glitch which prevents booting the real OS after unlocking SEDs.
+            # Symptom: Boot hangs when Grub hands over control to the Linux kernel. Seen on an HPE ML10Gen9 server.
+            # Cf. https://github.com/rear/rear/issues/2425
+            read -r field_name current_boot_number field_junk <<<"$(efibootmgr 2>/dev/null | grep 'BootCurrent:')"
```
Comment (jsmeix) [inline usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:208]: Only a side note from me who is not at all a UEFI expert
because you grep for `BootCurrent`:
You my have a look at my
https://github.com/rear/rear/issues/2276#issuecomment-559791942
and the subsequent comments from @gozora 
that there are various different ways how the EFI firmware can boot something
so that the `efibootmgr` output could be not reliable - as far as I understand things.
But of course for a workaround for a possinle EFI firmware glitch parsing
the `efibootmgr` output is at least better than nothing.

File: usr/share/rear/skel/default/etc/scripts/unlock-opal-disks [hunk 2]
```diff
@@ -201,6 +201,20 @@ while (( attempt < max_authentications )); do
...
+            if [[ "$current_boot_number" =~ ^[0-9]+$ ]]; then
+                # Set the current boot number as the boot entry for the next boot. This does not make much sense,
+                # except that we can be sure that we're setting a valid boot entry which we can then remove.
+                efibootmgr --bootnext "$current_boot_number" --quiet
+                # Remove the next boot entry. This should convince the firmware to use the boot order to determine
+                # the effective boot entry for the next boot.
+                efibootmgr --delete-bootnext --quiet
+            fi
```
Comment (jsmeix) [inline usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:216]: When $current_boot_number has not the expected form
would it make sense to `echo` some appropriate message to the user
to have him at least informed that things could go wrong here?

File: usr/share/rear/skel/default/etc/scripts/unlock-opal-disks (hunk 1 above)
Comment (OliverO2) [reply usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:208]: Yes, UEFI firmware can boot even without a proper boot entry present. `RAWDISK` output and other places within ReaR use the default bootloader path, which UEFI systems will discover automatically: https://github.com/rear/rear/blob/fb23c5d711af9ee505a9b03ea7324a098f90891d/usr/share/rear/output/RAWDISK/Linux-i386/270_create_grub2_efi_bootloader.sh#L103-L105

That does not mean that `efibootmgr` output is not reliable. Actually, it is. The tool does nothing but operate on the list of boot entries stored in the machine's NVRAM. `efibootmgr` does the same as your system's firmware setup with regard to boot entries and boot order.

File: usr/share/rear/skel/default/etc/scripts/unlock-opal-disks (hunk 2 above)
Comment (OliverO2) [reply usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:216]: I wanted the code to be completely silent even in case of errors as it is normally not required to work at all. So if `efibootmgr` spits out something unexpected, silent failure would not disturb or delay the boot process. This fix deals with a probably very rare condition and should not compromise a pleasant boot experience for others.

However, I just noticed that boot entry numbers can contain hex digits. If you happen to see more than 10 boot entries, and boot from entry `a` or above, the code would fail. I'll post another commit to account for that improbable case, too.

File: usr/share/rear/skel/default/etc/scripts/unlock-opal-disks (hunk 1 above)
Comment (jsmeix) [reply usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:208]: O.k. I rephrase it to:
"While the `efibootmgr` output is technically 100% reliable,
its meaning by normal humans interpretation is not."
;-)

Comment (OliverO2) [reply usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:208]: Well, yes, the entire UEFI thing is not that user friendly. It takes some time to get used to it... ;-)

File: usr/share/rear/skel/default/etc/scripts/unlock-opal-disks (hunk 2 above)
Comment (jsmeix) [reply usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:216]: FYI: It seems
https://github.com/rear/rear/issues/2276#issuecomment-559879836
shows an interesting example of boot entry numbers with hex digits.

Comment (OliverO2) [reply usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:216]: Yes, that one looks interesting.

If entries have piled up from testing, I usually use `efibootmgr` to clean up the mess. Not each system setup allows a user to really delete boot entries.

If a large number of boot entries is actually in use, I hope that they have excellent documentation - or better automation - in place to stay in control. ;-)

File: usr/share/rear/skel/default/etc/scripts/unlock-opal-disks [hunk 3]
```diff
@@ -201,6 +201,20 @@ while (( attempt < max_authentications )); do
         else
             display_message "$unlocked_device_count of $device_count disks unlocked, rebooting..."
         fi
+        if type -p efibootmgr &>/dev/null; then
```
Comment (gozora) [inline usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:204]: Just a shout question, why are you sending `type -p efibootmgr` to background using `&` ?

Comment (jsmeix) [reply usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:204]: @gozora 
`&>/dev/null` without a space between `&` and `>/dev/null`
is bash syntax shortcut to redirect both stdout and stderr to /dev/null

Comment (jsmeix) [reply usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:204]: SCNR:
"While the bash man page is technically 100% reliable,
its meaning by normal humans interpretation is not."
;-)

Comment (gozora) [reply usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:204]: That explains a lot ;-)
Thanks @jsmeix!

Comment (OliverO2) [reply usr/share/rear/skel/default/etc/scripts/unlock-opal-disks:204]: Thanks for reviewing and resolving. Faster than I could notice. ;-)

**Comment by jsmeix** [comment]
This one is obsoleted and superseded by https://github.com/rear/rear/pull/2455
